src/base/misc/disks.c

```diff
@@ -915,8 +918,17 @@ static void image_auto(struct disk *dp)
 	    filesize & (SECTOR_SIZE - 1) );
     }
     dp->num_secs = (filesize /* + SECTOR_SIZE - 1 */ ) / SECTOR_SIZE;
-    dp->heads = 255;
-    dp->sectors = 63;
+
+    /* Try to read the VBR of the first partition to learn spt and num heads */
```

Comment:
But will that work with the
multi-partition images?
Hmm, I wonder why vbr has
such info and mbr not...

Reply:
It will work in a sense that the number of sectors/track and number of heads shouldn't ever be different across partitions, but say if the first partition was non-DOS type then it won't have the required info and will fall back to the current hardcoded values. However I could scan all the partitions looking for the first DOS partition to read geometry from.
The VBR was developed first when there were no partitions (DOS 1.0), remember a volume is just like a floppy. The MBR came later (DOS 2.0) as a wrapper around multiple volumes, so I guess they didn't feel like duplicating data and just relied on the bios geometry to be able to find the start of the partition. But that's only a guess.

Reply:
> But you just said it only is used
with a non-mbr image. So how
can you "scan all partitions"?

I'm sorry to be confusing. I said the isfat32 variable is only used with non-mbr images. This is in a patch that fixes the geometry detection for MBR images that have no Dosemu header(that contains geometry). I will split up the patch series to make things more clear.

Reply:
OK I see.
So in both cases you need vbr,
then how about adding the vbr
struct to struct disk, instead of
isfat32?

Reply:
okay, yes I'll do that.